feat: sdk/trace: span processed metric for simple span processor

mahendrabishnoi2 opened the pull request:
Fixes #7004 

This PR adds support for experimental otel.sdk.processor.span.processed metric in simple span processor.
Definition of metric at: https://github.com/open-telemetry/semantic-conventions/blob/v1.36.0/docs/otel/sdk-metrics.md

Experimental metrics are behind a feature flag: `OTEL_GO_X_OBSERVABILITY`

<details>
<summary>Observability Implementation Checklist</summary>

## Observability Implementation Checklist
Based on the [project Observability guidelines](https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/CONTRIBUTING.md#observability), ensure the following are completed:

### Environment Variable Activation
* [x] Observability features are disabled by default
* [x] Features are activated through the `OTEL_GO_X_OBSERVABILITY` environment variable
* [x] Use consistent pattern with `x.Observability.Enabled()` check [^1]
* [x] Follow established experimental feature pattern [^2][^3]

[^1]: https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/observ/instrumentation.go#L101-L103
[^2]: https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/x/x.go
[^3]: https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/sdk/internal/x/x.go

### Encapsulation
* [x] Instrumentation is encapsulated within a dedicated `struct` (e.g., [`Instrumentation`](https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/observ/instrumentation.go#L86-L94))
* [x] Instrumentation is not mixed into the instrumented component
* [x] Instrumentation code is in its own file or package if complex/reused
* [x] Instrumentation setup doesn't bloat the main component code

### Initialization
* [x] Initialization is only done when observability is enabled
* [x] Setup is explicit and side-effect free
* [x] Return errors from initialization when appropriate
* [x] Use the global Meter provider (e.g., `otel.GetMeterProvider()`)
* [x] Include proper meter configuration with:
  * [x] Instrumentation package name is used for the Meter
  * [x] Instrumentation version (e.g. [`Version`](https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/observ/instrumentation.go#L40-L43))
  * [x] Schema URL (e.g. [`SchemaURL`](https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/observ/instrumentation.go#L36-L38))

### Performance
* [x] Little to no overhead when observability is disabled
* [x] Expensive operations are only executed when observability is enabled
* [x] When enabled, instrumentation code paths are optimized to reduce allocation/computation overhead

#### Attribute and Option Allocation Management
* [x] Use `sync.Pool` for attribute slices and options with dynamic attributes
* [x] Pool objects are properly reset before returning to pool
* [x] Pools are scoped for maximum efficiency while ensuring correctness

#### Caching
* [x] Static attribute sets known at compile time are pre-computed and cached
* [x] Common attribute combinations use lookup tables/maps

#### Benchmarking
* [x] Benchmarks provided for all instrumentation code
* [ ] Benchmark scenarios include both enabled and disabled observability
* [x] Benchmark results show impact on allocs/op, B/op, and ns/op (use `b.ReportAllocs()` in benchmarks)

### Error Handling and Robustness
* [x] Errors are reported back to caller when possible
* [x] Partial failures are handled gracefully
* [x] Use partially initialized components when available
* [x] Return errors to caller instead of only using `otel.Handle()`
* [x] Use `otel.Handle()` only when component cannot report error to user

### Context Propagation
* [x] Observability measurements receive the context from the function being measured (don't break context propagation by using `context.Background()`)

### Semantic Conventions Compliance
* [x] All metrics follow [OpenTelemetry Semantic Conventions](https://github.com/open-telemetry/semantic-conventions/blob/5ee549b1ce30fe11fcb9b7e3bd35ebfb363f467f/docs/otel/sdk-metrics.md)
* [x] Use the [`otelconv`](https://pkg.go.dev/go.opentelemetry.io/otel@v1.38.0/semconv/v1.37.0/otelconv) convenience package for metric semantic conventions
* [x] Component names follow semantic conventions
* [x] Use package path scope type as stable identifier for non-standard components
* [x] Component names are stable unique identifiers
* [x] Use global counter for uniqueness if necessary
* [x] Component ID counter is resettable for deterministic testing

### Testing
* [x] Use deterministic testing with isolated state
* [x] Restore previous state after tests (`t.Cleanup()`)
* [x] Isolate meter provider for testing
* [x] Use `t.Setenv()` for environment variable testing
* [x] Reset component ID counter for deterministic component names
* [x] Test order doesn't affect results
</details>

### Benchmarks

```console
> benchstat bmark.result
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/sdk/trace/internal/observ
cpu: Apple M1 Pro
                             │ bmark.result │
                             │    sec/op    │
SSP/SpanProcessed-8            146.7n ± 15%
SSP/SpanProcessedWithError-8   205.1n ±  3%
geomean                        173.5n

                             │ bmark.result │
                             │     B/op     │
SSP/SpanProcessed-8              280.0 ± 0%
SSP/SpanProcessedWithError-8     408.0 ± 0%
geomean                          338.0

                             │ bmark.result │
                             │  allocs/op   │
SSP/SpanProcessed-8              3.000 ± 0%
SSP/SpanProcessedWithError-8     3.000 ± 0%
geomean                          3.000
```

flc1125 commented on sdk/trace/internal/observ/simple_span_processor_test.go:
```diff
@@ -0,0 +1,126 @@
+	semconv "go.opentelemetry.io/otel/semconv/v1.37.0"
+)
+
+const sspComponentId = 0
```
```suggestion
const sspComponentID = 0
```

flc1125 commented on sdk/trace/simple_span_processor.go:
```diff
@@ -33,11 +38,26 @@ func NewSimpleSpanProcessor(exporter SpanExporter) SpanProcessor {
+	var err error
+	ssp.inst, err = observ.NewSSP(nextSimpleProcessorID())
+	if err != nil {
+		otel.Handle(err)
```
I think we can directly return the error here.

flc1125 replied:
No, please ignore this question. If adjusted in this way, it will not be backward compatible.

flc1125 commented on sdk/trace/simple_span_processor.go:
```diff
@@ -47,9 +67,16 @@ func (ssp *simpleSpanProcessor) OnEnd(s ReadOnlySpan) {
+			// Add the span to the context to ensure the metric is recorded
+			// with the correct span context.
+			ctx := trace.ContextWithSpanContext(context.Background(), s.SpanContext())
+			ssp.inst.SpanProcessed(ctx, 1, err)
```
I think there is a lack of reporting for situations beyond this part.

```go
if ssp.exporter != nil && s.SpanContext().TraceFlags().IsSampled() {
```

flc1125 replied:
ref: https://github.com/open-telemetry/semantic-conventions/blob/v1.36.0/docs/otel/sdk-metrics.md#metric-otelsdkprocessorspanprocessed

> For the SDK Simple and Batching Span Processor a span is considered to be processed already when it has been submitted to the exporter, not when the corresponding export call has finished.

- My understanding is: the exporter will perform self-observation on its own. If we only maintain the logic of `ExportSpans` here, then I believe the metrics here should be almost identical to those of the exporter.
- Based on this, I think it focuses more on the overall aspect, specifically the `OnEnd` part.

CC @MrAlias Please help confirm or explain the content here.

mahendrabishnoi2 replied:
@MrAlias please provide your inputs.

flc1125 commented:
We need to supplement the benchmark testing.

mahendrabishnoi2 commented:
> We need to supplement the benchmark testing.

```
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/sdk/trace/internal/observ
cpu: Apple M1 Pro
                             │  oldssp.txt   │                 newssp.txt                  │
                             │    sec/op     │     sec/op       vs base                    │
SSP/SpanProcessed-8            0.1114n ±  3%   144.3500n ± 12%  +129536.28% (p=0.000 n=10)
SSP/SpanProcessedWithError-8   0.1120n ± 19%   208.9500n ± 18%  +186545.82% (p=0.000 n=10)
geomean                        0.1116n            173.7n        +155450.86%

                             │ oldssp.txt │          newssp.txt          │
                             │    B/op    │    B/op     vs base          │
SSP/SpanProcessed-8            0.0 ± 0%     280.0 ± 0%  ? (p=0.000 n=10)
SSP/SpanProcessedWithError-8   0.0 ± 0%     408.0 ± 0%  ? (p=0.000 n=10)
geomean                                 ¹   338.0       ?
¹ summaries must be >0 to compute geomean

                             │  oldssp.txt  │          newssp.txt          │
                             │  allocs/op   │ allocs/op   vs base          │
SSP/SpanProcessed-8            0.000 ± 0%     3.000 ± 0%  ? (p=0.000 n=10)
SSP/SpanProcessedWithError-8   0.000 ± 0%     3.000 ± 0%  ? (p=0.000 n=10)
geomean                                   ¹   3.000       ?
¹ summaries must be >0 to compute geomean

```
1. Disable observability and run `go test -bench='BenchmarkSSP' -count=10 > newssp.txt`.
2. Enable observability and run `go test -bench='BenchmarkSSP' -count=10 > oldssp.txt`.
3. Using benchstat to generate above numbers `benchstat oldssp.txt newssp.txt`.

Note: For disabling case, I modified BenchmarkSSP test case to set env var as `false` and added below code block in benchmark loop.
```go
if ssp == nil {
	continue
}
```

I'm not sure if this is the right way or not. Looking at benchmarks on other PRs, I can see that the number for old and new are supposed to be equivalent. What am I missing?

flc1125 commented:
> │  oldssp.txt   │                 newssp.txt                  │
>                              │    sec/op     │     sec/op       vs base                    │
> SSP/SpanProcessed-8            0.1114n ±  3%   144.3500n ± 12%  +129536.28% (p=0.000 n=10)
> SSP/SpanProcessedWithError-8   0.1120n ± 19%   208.9500n ± 18%  +186545.82% (p=0.000 n=10)

This data seems a bit exaggerated. Can we ensure it's correct? Or is there an issue with the program somewhere?

mahendrabishnoi2 commented:
> > │  oldssp.txt   │                 newssp.txt                  │
> > │    sec/op     │     sec/op       vs base                    │
> > SSP/SpanProcessed-8            0.1114n ±  3%   144.3500n ± 12%  +129536.28% (p=0.000 n=10)
> > SSP/SpanProcessedWithError-8   0.1120n ± 19%   208.9500n ± 18%  +186545.82% (p=0.000 n=10)
> 
> This data seems a bit exaggerated. Can we ensure it's correct? Or is there an issue with the program somewhere?

That's what I thought. So, I want to know if the steps I took were right or not?
Numbers seem right to me considering what I'm benchmarking. In one case when observability is disabled, nothing would be executed.

MrAlias commented:
> > > │  oldssp.txt   │                 newssp.txt                  │
> > > │    sec/op     │     sec/op       vs base                    │
> > > SSP/SpanProcessed-8            0.1114n ±  3%   144.3500n ± 12%  +129536.28% (p=0.000 n=10)
> > > SSP/SpanProcessedWithError-8   0.1120n ± 19%   208.9500n ± 18%  +186545.82% (p=0.000 n=10)
> > 
> > 
> > This data seems a bit exaggerated. Can we ensure it's correct? Or is there an issue with the program somewhere?
> 
> That's what I thought. So, I want to know if the steps I took were right or not? Numbers seem right to me considering what I'm benchmarking. In one case when observability is disabled, nothing would be executed.

How did you generate the `oldssp.txt`? Are you making a valid comparison?

mahendrabishnoi2 commented:
> > > > │  oldssp.txt   │                 newssp.txt                  │
> > > > │    sec/op     │     sec/op       vs base                    │
> > > > SSP/SpanProcessed-8            0.1114n ±  3%   144.3500n ± 12%  +129536.28% (p=0.000 n=10)
> > > > SSP/SpanProcessedWithError-8   0.1120n ± 19%   208.9500n ± 18%  +186545.82% (p=0.000 n=10)
> > > 
> > > 
> > > This data seems a bit exaggerated. Can we ensure it's correct? Or is there an issue with the program somewhere?
> > 
> > 
> > That's what I thought. So, I want to know if the steps I took were right or not? Numbers seem right to me considering what I'm benchmarking. In one case when observability is disabled, nothing would be executed.
> 
> How did you generate the `oldssp.txt`? Are you making a valid comparison?

Details of how I generated the file is mentioned in https://github.com/open-telemetry/opentelemetry-go/pull/7374#issuecomment-3378748849.

I simply disabled observability and used `go test -bench='BenchmarkSSP' -count=10 > oldssp.txt` command.

mahendrabishnoi2 commented:
> Your methodology is unclear. Changing the environment variable to disable observability will result in either errors or panics.
> 
> Like I mentioned before, it is not clear how you `oldssp.txt`. If you completely gutted your testing to just measure no-ops I would expect to see results like you are posting. I don't see much utility in measurements like this. Instead, I would just provide benchstat results for added benchmark without any "old" version.

Makes sense. Yes, one of the test (oldssp.txt) was basically no-op as observability was disabled. I looked at other PRs and those had comparisons so I thought of adding that. Now that I understand this better, I believe those numbers were generated by comparing against main branch and not by disabling feature (no-op).
I'll add benchstat results in a separate comment.

Thank you for your help on this. Appreciate it.

mahendrabishnoi2 commented:
benchmarks:
```
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/sdk/trace/internal/observ
cpu: Apple M1 Pro
BenchmarkSSP/SpanProcessed-8         	 8767838	       136.8 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 8576318	       146.1 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 8446096	       188.4 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 6529568	       169.2 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 7442804	       146.8 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 8366742	       144.4 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 7617855	       146.6 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 8387133	       146.5 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 8051401	       147.1 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessed-8         	 8068048	       147.8 ns/op	     280 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5906968	       201.4 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5790499	       205.2 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5860546	       199.3 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5978707	       210.6 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5844382	       205.9 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5869333	       199.7 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 6049657	       200.2 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5929299	       208.5 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5690490	       206.9 ns/op	     408 B/op	       3 allocs/op
BenchmarkSSP/SpanProcessedWithError-8         	 5836519	       205.0 ns/op	     408 B/op	       3 allocs/op
```

MrAlias commented:
@mahendrabishnoi2 based your [results](https://github.com/open-telemetry/opentelemetry-go/pull/7374#issuecomment-3381995390):

```console
> benchstat bmark.result
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/sdk/trace/internal/observ
cpu: Apple M1 Pro
                             │ bmark.result │
                             │    sec/op    │
SSP/SpanProcessed-8            146.7n ± 15%
SSP/SpanProcessedWithError-8   205.1n ±  3%
geomean                        173.5n

                             │ bmark.result │
                             │     B/op     │
SSP/SpanProcessed-8              280.0 ± 0%
SSP/SpanProcessedWithError-8     408.0 ± 0%
geomean                          338.0

                             │ bmark.result │
                             │  allocs/op   │
SSP/SpanProcessed-8              3.000 ± 0%
SSP/SpanProcessedWithError-8     3.000 ± 0%
geomean                          3.000
```

MrAlias commented on sdk/trace/internal/observ/simple_span_processor.go:
```diff
@@ -0,0 +1,78 @@
+	if err != nil {
+		*attrs = append(*attrs, semconv.ErrorType(err))
+	}
+	*attrs = append(*attrs,
+		ssp.componentNameAttr,
+		ssp.spansProcessedCounter.AttrComponentType(otelconv.ComponentTypeSimpleSpanProcessor))
+	ssp.spansProcessedCounter.Add(ctx, count, *attrs...)
```
These attributes are known at the creation of the `SSP`. They should be made into an option at the start so the allocations added here are not needed when there is no error.

Please take a look at the other instrumentation.

MrAlias commented on sdk/trace/internal/observ/simple_span_processor.go:
```diff
@@ -0,0 +1,78 @@
+	},
+}
+
+type SSP struct {
```
Missing documentation.

MrAlias commented on sdk/trace/internal/observ/simple_span_processor.go:
```diff
@@ -0,0 +1,78 @@
+	return semconv.OTelComponentName(name)
+}
+
+func NewSSP(id int64) (*SSP, error) {
```
Missing documentation.

MrAlias commented on sdk/trace/simple_span_processor_test.go:
```diff
@@ -168,3 +191,140 @@ func TestSimpleSpanProcessorShutdownHonorsContextCancel(t *testing.T) {
 	}
 }
+
+func TestSimpleSpanProcessorSelfObservability(t *testing.T) {
```
```suggestion
func TestSimpleSpanProcessorObservability(t *testing.T) {
```

MrAlias commented on sdk/trace/simple_span_processor_test.go:
```diff
@@ -168,3 +191,140 @@ func TestSimpleSpanProcessorShutdownHonorsContextCancel(t *testing.T) {
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Setenv("OTEL_GO_X_SELF_OBSERVABILITY", strconv.FormatBool(test.enabled))
```
```suggestion
			t.Setenv("OTEL_GO_X_OBSERVABILITY", strconv.FormatBool(test.enabled))
```

MrAlias reviewed (commented):
Overall, this looks good. The memory management of attributes during the measurement needs to be optimized though.

MrAlias commented:
@mahendrabishnoi2 this looks ready to merge. Please take a look at the conflicts and resolve them. `sdk/trace/internal/x/README.md` was moved to `sdk/internal/x/README.md`.
